emloop/main_loop.py

```diff
@@ -98,6 +98,24 @@ def __init__(self,   # pylint: disable=too-many-arguments
 
         super().__init__()
 
+    def __enter__(self):
+        """Calls before_training() for all hooks."""
+        self.hooks_before_training()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Calls after_training() for all hooks."""
+        self.hooks_after_training()
+
+    def hooks_before_training(self):
```

Comment:
I would name this `trigger_before_training` (similarly for the following method).